File: python/25_linked_list_cycle.py
#!/usr/bin/env python3
"""
LeetCode #141: 环形链表
题目: 判断链表中是否存在环。
解法: 快慢指针（Floyd 判圈） - 快指针每次走两步，慢指针每次走一步。
      若存在环，两者必然在环内相遇。时间 O(n)，空间 O(1)。
"""


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def hasCycle(head: ListNode) -> bool:
    """
    快慢指针: fast 走两步，slow 走一步。
    若链表有环，fast 和 slow 必然在环内相遇。
    """
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
    return False


if __name__ == "__main__":
    # 构建环: 1->2->3->4->2
    n1 = ListNode(1); n2 = ListNode(2); n3 = ListNode(3); n4 = ListNode(4)
    n1.next = n2; n2.next = n3; n3.next = n4; n4.next = n2
    print(hasCycle(n1))  # True
    # 无环
    n1.next = n2; n2.next = n3; n3.next = n4; n4.next = None
    print(hasCycle(n1))  # False

File: cpp/25_linked_list_cycle.cpp
/**
 * LeetCode #141: 环形链表
 * 题目: 判断链表中是否存在环。
 * 解法: 快慢指针（Floyd 判圈）- 快指针每次走两步，慢指针每次走一步。
 *       若存在环，两者必然在环内相遇。O(n)，O(1)。
 */

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    bool hasCycle(ListNode* head) {
        ListNode *slow = head, *fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) return true;
        }
        return false;
    }
};

int main() {
    Solution sol;
    ListNode *n1 = new ListNode(1), *n2 = new ListNode(2),
             *n3 = new ListNode(3), *n4 = new ListNode(4);
    n1->next = n2; n2->next = n3; n3->next = n4; n4->next = n2; // 环形
    cout << sol.hasCycle(n1) << endl; // 1 (true)
    n4->next = nullptr;
    cout << sol.hasCycle(n1) << endl; // 0 (false)
    return 0;
}
